// ng-http-loader/esm2015/lib/ng-http-loader.module.js
/*
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
var NgHttpLoaderModule_1;
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgHttpLoaderComponent } from './components/ng-http-loader.component';
import { PendingRequestsInterceptorProvider } from './services/pending-requests-interceptor.service';
import { SPINKIT_COMPONENTS } from './spinkits';
let NgHttpLoaderModule = NgHttpLoaderModule_1 = class NgHttpLoaderModule {
    static forRoot() {
        return {
            ngModule: NgHttpLoaderModule_1,
            providers: [
                PendingRequestsInterceptorProvider,
            ],
        };
    }
};
NgHttpLoaderModule = NgHttpLoaderModule_1 = __decorate([
    NgModule({
        declarations: [
            NgHttpLoaderComponent,
            ...SPINKIT_COMPONENTS,
        ],
        imports: [
            CommonModule,
        ],
        exports: [
            NgHttpLoaderComponent,
            ...SPINKIT_COMPONENTS,
        ],
    })
], NgHttpLoaderModule);
export { NgHttpLoaderModule };
