// @ncstate/sat-popover/esm5/popover/popover.errors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { VALID_HORIZ_ALIGN, VALID_VERT_ALIGN, VALID_SCROLL } from './types';
/**
 * @return {?}
 */
export function getUnanchoredPopoverError() {
    return Error('SatPopover does not have an anchor.');
}
/**
 * @return {?}
 */
export function getInvalidPopoverAnchorError() {
    return Error('SatPopover#anchor must be an instance of SatPopoverAnchor, ElementRef, or HTMLElement.');
}
/**
 * @return {?}
 */
export function getInvalidPopoverError() {
    return Error('SatPopoverAnchor#satPopoverAnchor must be an instance of SatPopover.');
}
/**
 * @return {?}
 */
export function getInvalidSatPopoverAnchorError() {
    return Error("SatPopoverAnchor must be associated with a " +
        "SatPopover component. " +
        "Examples: <sat-popover [anchor]=\"satPopoverAnchorTemplateRef\"> or " +
        "<button satPopoverAnchor [satPopoverAnchor]=\"satPopoverTemplateRef\">");
}
/**
 * @param {?} alignment
 * @return {?}
 */
export function getInvalidHorizontalAlignError(alignment) {
    return Error(generateGenericError('horizontalAlign/xAlign', alignment, VALID_HORIZ_ALIGN));
}
/**
 * @param {?} alignment
 * @return {?}
 */
export function getInvalidVerticalAlignError(alignment) {
    return Error(generateGenericError('verticalAlign/yAlign', alignment, VALID_VERT_ALIGN));
}
/**
 * @param {?} strategy
 * @return {?}
 */
export function getInvalidScrollStrategyError(strategy) {
    return Error(generateGenericError('scrollStrategy', strategy, VALID_SCROLL));
}
/**
 * @param {?} apiName
 * @param {?} invalid
 * @param {?} valid
 * @return {?}
 */
function generateGenericError(apiName, invalid, valid) {
    return "Invalid " + apiName + ": '" + invalid + "'. Valid options are " + valid.map((/**
     * @param {?} v
     * @return {?}
     */
    function (v) { return "'" + v + "'"; })).join(', ') + ".";
}
