// @ncstate/sat-popover/esm5/popover/popover-hover.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input } from '@angular/core';
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { of, Subject } from 'rxjs';
import { delay, switchMap, takeUntil } from 'rxjs/operators';
import { SatPopoverAnchor } from './popover.component';
var SatPopoverHoverDirective = /** @class */ (function () {
    function SatPopoverHoverDirective(anchor) {
        this.anchor = anchor;
        this._satPopoverHover = 0;
        /**
         * Emits when the directive is destroyed.
         */
        this._onDestroy = new Subject();
        /**
         * Emits when the user's mouse enters the element.
         */
        this._onMouseEnter = new Subject();
        /**
         * Emits when the user's mouse leaves the element.
         */
        this._onMouseLeave = new Subject();
    }
    Object.defineProperty(SatPopoverHoverDirective.prototype, "satPopoverHover", {
        /**
         * Amount of time to delay (ms) after hovering starts before
         * the popover opens. Defaults to 0ms.
         */
        get: /**
         * Amount of time to delay (ms) after hovering starts before
         * the popover opens. Defaults to 0ms.
         * @return {?}
         */
        function () { return this._satPopoverHover; },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._satPopoverHover = coerceNumberProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SatPopoverHoverDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Whenever the user hovers this host element, delay the configured
        // amount of time and open the popover. Terminate if the mouse leaves
        // the host element before the delay is complete.
        this._onMouseEnter
            .pipe(switchMap((/**
         * @return {?}
         */
        function () {
            return of(null).pipe(delay(_this._satPopoverHover || 0), takeUntil(_this._onMouseLeave));
        })), takeUntil(this._onDestroy))
            .subscribe((/**
         * @return {?}
         */
        function () { return _this.anchor.popover.open(); }));
    };
    /**
     * @return {?}
     */
    SatPopoverHoverDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._onDestroy.next();
        this._onDestroy.complete();
    };
    /**
     * @return {?}
     */
    SatPopoverHoverDirective.prototype.showPopover = /**
     * @return {?}
     */
    function () {
        this._onMouseEnter.next();
    };
    /**
     * @return {?}
     */
    SatPopoverHoverDirective.prototype.closePopover = /**
     * @return {?}
     */
    function () {
        this._onMouseLeave.next();
        this.anchor.popover.close();
    };
    SatPopoverHoverDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[satPopoverHover]'
                },] }
    ];
    /** @nocollapse */
    SatPopoverHoverDirective.ctorParameters = function () { return [
        { type: SatPopoverAnchor }
    ]; };
    SatPopoverHoverDirective.propDecorators = {
        satPopoverHover: [{ type: Input }],
        showPopover: [{ type: HostListener, args: ['mouseenter',] }],
        closePopover: [{ type: HostListener, args: ['mouseleave',] }]
    };
    return SatPopoverHoverDirective;
}());
export { SatPopoverHoverDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SatPopoverHoverDirective.prototype._satPopoverHover;
    /**
     * Emits when the directive is destroyed.
     * @type {?}
     * @private
     */
    SatPopoverHoverDirective.prototype._onDestroy;
    /**
     * Emits when the user's mouse enters the element.
     * @type {?}
     * @private
     */
    SatPopoverHoverDirective.prototype._onMouseEnter;
    /**
     * Emits when the user's mouse leaves the element.
     * @type {?}
     * @private
     */
    SatPopoverHoverDirective.prototype._onMouseLeave;
    /** @type {?} */
    SatPopoverHoverDirective.prototype.anchor;
}
